File: lib/models/od_request.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:hive/hive.dart';

part 'od_request.g.dart';

@HiveType(typeId: 1)
@JsonSerializable()
class ODRequest {
  @HiveField(0)
  final String id;
  
  @HiveField(1)
  final String studentId;
  
  @HiveField(2)
  final String studentName;
  
  @HiveField(3)
  final String registerNumber;
  
  @HiveField(4)
  final DateTime date;
  
  @HiveField(5)
  final List<int> periods;
  
  @HiveField(6)
  final String reason;
  
  @HiveField(7)
  final String status; // 'pending', 'approved', 'rejected'
  
  @HiveField(8)
  final String? attachmentUrl;
  
  @HiveField(9)
  final DateTime createdAt;
  
  @HiveField(10)
  final DateTime? approvedAt;
  
  @HiveField(11)
  final String? approvedBy;
  
  @HiveField(12)
  final String? rejectionReason;

  @HiveField(13)
  final String? staffId;

  const ODRequest({
    required this.id,
    required this.studentId,
    required this.studentName,
    required this.registerNumber,
    required this.date,
    required this.periods,
    required this.reason,
    required this.status,
    this.attachmentUrl,
    required this.createdAt,
    this.approvedAt,
    this.approvedBy,
    this.rejectionReason,
    this.staffId,
  });

  factory ODRequest.fromJson(Map<String, dynamic> json) => _$ODRequestFromJson(json);
  
  Map<String, dynamic> toJson() => _$ODRequestToJson(this);
  
  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';
}

File: backend/app/models/od_request.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base
from app.core.enums import ODStatus

class ODRequest(Base):
    __tablename__ = "od_requests"

    id = Column(Integer, primary_key=True, index=True)
    
    # Relationships
    student_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    student = relationship("User", foreign_keys=[student_id], backref="od_requests")
    
    # Request Details
    register_number = Column(String, index=True) # Redundant but useful for quick lookup
    student_name = Column(String) # Redundant but useful snapshot
    date = Column(DateTime(timezone=True), nullable=False)
    periods = Column(JSON, nullable=False) # List of period numbers [1, 2, 3]
    reason = Column(Text, nullable=False)
    attachment_url = Column(String, nullable=True)
    
    # Status & Approval
    status = Column(String, default=ODStatus.PENDING, index=True) # pending, approved, rejected
    
    approved_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    approved_by = relationship("User", foreign_keys=[approved_by_id])
    
    approved_at = Column(DateTime(timezone=True), nullable=True)
    rejection_reason = Column(Text, nullable=True)
    
    # Audit
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
